blocklib/picoscope4000a/picoscope4000a.hpp

```diff
@@ -0,0 +1,73 @@
+#ifndef GR_PICOSCOPE4000A_PICOSCOPE4000A_HPP
+#define GR_PICOSCOPE4000A_PICOSCOPE4000A_HPP
+
+#include <picoscope.hpp>
+
+namespace gr::picoscope4000a {
+
+struct Picoscope4000a : public gr::picoscope::Picoscope<Picoscope4000a> {
+    using AnalogPort = fair::graph::PortOut<float>;
+
+    AnalogPort values0;
+    AnalogPort errors0;
+    AnalogPort values1;
+    AnalogPort errors1;
+    AnalogPort values2;
+    AnalogPort errors2;
+    AnalogPort values3;
+    AnalogPort errors3;
+    AnalogPort values4;
+    AnalogPort errors4;
+    AnalogPort values5;
+    AnalogPort errors5;
+    AnalogPort values6;
+    AnalogPort errors6;
+    AnalogPort values7;
+    AnalogPort errors7;
+
+    auto
+    channel_outputs() {
+        return std::array<std::pair<AnalogPort &, AnalogPort &>, 8>{
+            { { values0, errors0 }, { values1, errors1 }, { values2, errors2 }, { values3, errors3 }, { values4, errors4 }, { values5, errors5 }, { values6, errors6 }, { values7, errors7 } }
+        };
+    }
+
+    fair::graph::work_return_t
+    work(std::size_t requested_work = 0) {
+        std::ignore = requested_work;
+        return this->work_impl();
+    }
+
+    std::error_code
+                           set_buffers(size_t samples, uint32_t block_number);
+
+    static constexpr float DRIVER_VERTICAL_PRECISION = 0.01f;
+
+    static std::optional<std::size_t>
+    driver_channel_id_to_index(std::string_view id);
```

Comment:
That's not possible without clashing include paths, as libps4000a/ps4000api.h etc. directly include header files in their directory, like `#include <PicoCallback.h>`. Then a random header would be used, whatever comes first in the `-I` list, once you want to instantiate more than one PS subtype block. That's why all code that uses libpsX includes is in the .cc currently.

Reply:
There are no collisions: 
 * `PicoCallback.h` is only available for the PS5000 and PS6000 models. 
 * Each model has it's own header sub-directory, the include structure is for picoscope/include
 ```text
 include/
├── libpl1000
│   ├── PicoStatus.h
│   └── pl1000Api.h
├── libps2000
│   └── ps2000.h
├── libps2000a
│   ├── PicoStatus.h
│   └── ps2000aApi.h
├── libps3000
│   └── ps3000.h
├── libps3000a
│   ├── PicoStatus.h
│   └── ps3000aApi.h
├── libps4000
│   ├── PicoStatus.h
│   └── ps4000Api.h
├── libps4000a
│   ├── PicoConnectProbes.h
│   ├── PicoStatus.h
│   └── ps4000aApi.h
├── libps5000
│   ├── PicoStatus.h
│   └── ps5000Api.h
├── libps5000a
│   ├── PicoCallback.h
│   ├── PicoConnectProbes.h
│   ├── PicoDeviceEnums.h
│   ├── PicoDeviceStructs.h
│   ├── PicoStatus.h
│   ├── PicoVersion.h
│   └── ps5000aApi.h
├── libps6000
│   ├── PicoStatus.h
│   └── ps6000Api.h
├── libps6000a
│   ├── PicoCallback.h
│   ├── PicoConnectProbes.h
│   ├── PicoDeviceDefinitionsExperimental.h
│   ├── PicoDeviceEnums.h
│   ├── PicoDeviceStructs.h
│   ├── PicoStatus.h
│   ├── PicoVersion.h
│   ├── ps6000aApiExperimental.h
│   └── ps6000aApi.h
└── libusbdrdaq
    ├── PicoStatus.h
    └── usbDrDaqApi.h

 ```
 
 using 
 ```bash
 while read -r file; do
    locations=$(find include/ -type f -name "$file")
    prev=""
    for loc in $locations; do
        if [[ ! -z "$prev" ]]; then
            if ! cmp -s "$prev" "$loc"; then
                echo "Difference found between $prev and $loc"
            else
                echo "Contents are identical for $prev and $loc"
            fi
        fi
        prev=$loc
    done
done < duplicates.txt
 ```
 and
 ```bash
 while read -r file; do
    locations=$(find include/ -type f -name "$file")
    prev=""
    for loc in $locations; do
        if [[ ! -z "$prev" ]]; then
            if ! cmp -s "$prev" "$loc"; then
                echo "Difference found between $prev and $loc"
            else
                echo "Contents are identical for $prev and $loc"
            fi
        fi
        prev=$loc
    done
done < duplicates.txt
 ```
 you can see that all the duplicates are identical.
 
In summary: the Picotech guys made sure that you can use different models in an application/compilation unit w/o name or definition clashes. Also, there is always the option of declaring the models as different GR plugins which would put them into different compilation units. If we (ever) encounter problems with linkage, macro, or naming conflicts we could reiterate. Let's KISS and inline now. Easier to refactor, review and build upon.

Reply:
Yeah, ps4000a also includes files outside of its folder, which I considered a bug ("PicoCallback.h", "PicoConnectProbes.h", "PicoVersion.h"). I took the duplication as hint that these subfolders are meant to be self-contained and any files being identical is just coincidental, because they didn't diverge yet.
We can assume the files to be identical for now and inline everything. 
Maybe fear was that the headers diverge without us noticing and creating subtle bugs. Maybe we add a script like the above to the CI that warns us should that happen.